src/Psalm/Internal/DataFlow/TaintSource.php

```diff
@@ -9,4 +9,14 @@
  */
 final class TaintSource extends DataFlowNode
 {
+    public static function fromNode(DataFlowNode $node): self
+    {
+        return new self(
+            $node->id,
```

Comment:
Hi, looking through the taint system now, shouldn't this be `$node->unspecialized_id ?? $node->id` to avoid re-specialization of already specalized nodes?

Reply:
Honestly, I have no idea... maybe the impact can be when changing and checking if all tests still pass - at least if there are any specialization taint tests.

Reply:
In general my feedback to the taint system is, that it's really hard to overview and internal documentation is extremely short. In sum there are unfortunately only a few comments or docblocks in the codebase with explanations which makes it very hard to get into all the Analyzer classes and their (taint) handling.